src/authn/index.js

```diff
@@ -514,6 +619,73 @@ async function verifyToken(token, use, client = COGNITO_CLIENT_ID) {
   return claims;
 }
 
+
+/**
+ * Exchanges `refreshToken` for a new `idToken` and `accessToken`.
+ *
+ * @param {String} refreshToken
+ * @returns {Object} tokens
+ * @returns {String} tokens.idToken
+ * @returns {String} tokens.accessToken
+ * @throws {AuthnRefreshTokenInvalid} If the `refreshToken` is invalid,
+ *   expired, revoked, or was issued to another client.  Existing tokens should
+ *   be discarded if this error occurs as there is ~no way to recover from it
+ *   except by initiating a new login.
+ */
+async function renewTokens(refreshToken) {
+  const response = await fetch(`${COGNITO_BASE_URL}/oauth2/token`, {
+    method: "POST",
+    body: new URLSearchParams([
+      ["grant_type", "refresh_token"],
+      ["client_id", COGNITO_CLIENT_ID],
+      ["refresh_token", refreshToken],
+    ]),
+  });
+
+  const body = await response.text();
+
+  switch (response.status) {
+    case 200:
+      break;
+
+    case 400:
+      /* In the context of grant_type=refresh_token (above), the invalid_grant
+       * error means¹:
+       *
+       *   [the] refresh token is invalid, expired, revoked, does not match the
+       *   redirection URI used in the authorization request, or was issued to
+       *   another client.
+       *
+       * as described by the OAuth 2.0 spec¹, which is referenced by the error
+       * response section for the TOKEN endpoint² in the OpenID Connect Core
+       * 1.0 spec, which is what Cognito implements here.
+       *
+       * All other 400 Bad Request errors are programming/service issues not
+       * related to the refresh token itself, so fallthru to the default
+       * generic error.
+       *
+       * ¹ RFC 6749 § 5.2: https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
+       * ² https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.3.1.3.4
```

Comment:
Thank you for these links! They were very helpful for getting to docs for reading to understand what's going on in this PR 😄

Reply:
Oh good! I included them because it took some time to track them down, and I wanted an easier reference later. :-)